user-message-api

maxwell912 opened the pull request:
Поправил баг с запросом

user-message-api
Создаю таблицы при старте
Поправил фк сообщений

user-message-api
Роут для создания чата

user-message-api
Создал модели для юзеров, чатов и сообщений
Создал основные пути для работы с ними

Commit:
user-message-api
Поправил баг с запросом

user-message-api
Создаю таблицы при старте
Поправил фк сообщений

user-message-api
Роут для создания чата

user-message-api
Создал модели для юзеров, чатов и сообщений
Создал основные пути для работы с ними

SofiSheromova commented on server/controllers/messages.ts:
```diff
@@ -0,0 +1,35 @@
+        res.json(messages.reverse())
+    )
+        // eslint-disable-next-line no-console
+        .catch(err => console.log(err));
```
Если произошла ошибка, то отправь обратно сообщение об ошибке с соответствующий кодом
И используй console.error(), тогда не нужно будет писать комментарии для eslint

SofiSheromova replied:
что-то типа
 .catch(e => res.status(400).json({error: {code: 400, message: e.toString()}));

maxwell912 replied:
попраил

SofiSheromova commented on server/controllers/messages.ts:
```diff
@@ -0,0 +1,35 @@
+}, res: Response): void {
+    messageQuery.findAll({
+        order: [
+            ['createdAt', 'DESC']
```
А зачем DESC? Кажется отсортировать по созданию уже достаточно

maxwell912 replied:
Чтобы запрашивая первую страницу ты получал последний n сообщений 
Это вроде удобно

SofiSheromova replied:
А, тогда я не совсем правильно поняла в первый раз. Получается тут происходит сортировка по времени от нового к старому, так?

Gleb1nho replied:
Да, это работает так

SofiSheromova commented on server/controllers/messages.ts:
```diff
@@ -0,0 +1,35 @@
+        order: [
+            ['createdAt', 'DESC']
+        ],
+        offset: (page - 1) * count,
+        limit: count,
```
Идея хорошая, есть пара нюансов
Как на фронте понять, что он прочитал все сообщения? Отправляй с сообщениями информацию о их количестве
Чаще всего требуется получить самые последние сообщения, может странички считать в другую сторону? Тогда и фронту не придётся делать 2 запроса, чтобы отобразить хоть что-то, а только один 
И стоит установить дефолтное количество сообщений и номер страницы

maxwell912 replied:
У меня сообщения сортируютсчя по созданию в порядке убывания так что первая страница это последние n сообщний

maxwell912 replied:
Кол-во сообщений добавил

SofiSheromova commented on server/controllers/messages.ts:
```diff
@@ -0,0 +1,35 @@
+import { Response } from 'express';
+import { message as messageQuery } from '../models/message';
+
+export function list({ params: { id }, body: { page, count } }: {
```
id -> chatId (не сразу понятно что за id требуется)

maxwell912 replied:
поправил

SofiSheromova commented on server/controllers/messages.ts:
```diff
@@ -0,0 +1,35 @@
+        .catch(err => console.log(err));
+}
+
+export function create({ params: { id }, body: { userId, message } }: {
```
id -> chatId и в новой версии (которая сейчас в develop) было так, что в req есть поле user, в котором хранится текущий авторизированный пользователь
но мне честно пока сложно это комментировать, потому что сама не до конца разобралась ахпхах

maxwell912 replied:
поправил

SofiSheromova commented on server/controllers/messages.ts:
```diff
@@ -0,0 +1,35 @@
+        value: message
+    })
+        .then(r => res.status(200).json(r.toJSON()))
+        .catch(e => res.status(400).json(e.toString()));
```
добавь ещё console.error, пусть пока так отображается

SofiSheromova replied:
и  чтобы везде был один формат ошибки, тут тоже сделать
.catch(e => res.status(400).json({error: {code: 400, message: e.toString()}));

maxwell912 replied:
Если уже у ответа ставим статус 400 то зачем его вставлять еще и в боди?

SofiSheromova replied:
code и message это такой стандартный набор, когда передаёшь объект ошибки. Может потому что на фронте не очень удобно получать код 🤔 
Я не задумывалась об этом раньше

SofiSheromova replied:
в целом и так норм
но просто принято как-то

SofiSheromova commented on server/models/chat.ts:
```diff
@@ -0,0 +1,25 @@
+    // Модель чата
+    // id - SERIAL PRIMARY KEY
+    // createdAt - DATETIME время создания сообщения
+    name: {
```
Видимо ты решил использовать схему, что в чате могут быть несколько пользователей 
Может не будем усложнять? 
Ну или если усложнять, то надо ещё хранить людей в чате
Но я за то, чтоб не усложнять

SofiSheromova replied:
Я подумала ещё раз
У нас чат в задаче только с одним человеком. Мы видим список контактов и, нажимая на контакт, открывается чат
Поэтому можно не хранить название чата. Но должна быть возможность получить id чата по id пользователей

maxwell912 replied:
Заменил таблицу на список логинов пользователей

SofiSheromova commented on server/models/chatToUser.ts:
```diff
@@ -0,0 +1,32 @@
+        allowNull: false,
+        comment: 'Ид чата, в котором состоит пользователь'
+    },
+    userId: {
```
Типа отдельная таблица на то, чтоб хранить связь чатов и пользователей? В целом, если считаешь, что так удобней, то ок

maxwell912 replied:
Убрал

SofiSheromova commented on server/models/user.ts:
```diff
@@ -1,16 +1,49 @@
+        type: STRING,
+        comment: 'Ид пользователя на гитхабе'
+    },
+    firstName: {
+        type: STRING,
+        comment: 'Имя пользователя'
+    },
+    lastName: {
+        type: STRING,
+        comment: 'Фамилия пользователя'
+    },
+    middleName: {
+        type: STRING,
+        comment: 'Отчество пользователя'
+    },
+    email: {
+        type: STRING,
+        allowNull: false,
+        comment: 'Почтовый адрес пользователя'
+    },
+    phone: {
+        type: STRING,
+        comment: 'Телефонный номер пользователя'
+    },
+    avatar: {
+        type: STRING,
+        comment: 'Аватарка пользователя'
+    },
+    createdAt: {
+        type: DATE,
+        defaultValue: Sequelize.NOW
```
если мы хотим всё это хранить, то при создании пользователя нужно подтягивать с гитхаба нужные данные и сохранять их в базу

Gleb1nho replied:
Уже делается в новой авторизации, всё что можно нужное взять из апишки, берётся:)

SofiSheromova reviewed (commented):
controllers/user.ts тоже надо переделать, как минимум блок catch

SofiSheromova commented:
а ещё мы обновили develop, возможно тебе лучше обновить и свою ветку

maxwell912 committed:
Merge remote-tracking branch 'origin/develop' into user-message-api

# Conflicts:
#	server/routes.ts

maxwell912 committed:
user-message-api
Изменил тип хранения юзеров в чатах
Добавил обработку ошибок

maxwell912 committed:
Merge remote-tracking branch 'origin/prod-db-auth' into user-message-api

# Conflicts:
#	server/models/chat.ts
#	server/models/user.ts

maxwell912 committed:
user-message-api
Поменял названия моделей

maxwell912 committed:
user-message-api
Использую ники гитхаба как ид

maxwell912 committed:
user-message-api
Переименовал поле

maxwell912 committed:
user-message-api
Переименовал поле

maxwell912 committed:
Merge branch 'develop' into user-message-api

# Conflicts:
#	server/controllers/auth.ts
#	server/controllers/users.ts
#	server/models/chat.ts
#	server/models/message.ts
#	server/models/user.ts

SofiSheromova commented:
Кажется, что всё ок
Пока не попробую с этим поработать сложно сказать, но на вид всё есть
Только поправить конфликты

SofiSheromova commented on server/routes.ts:
```diff
@@ -14,10 +16,13 @@ export default (server: Application, passport: PassportStatic, nextApp: NextServ
 
-    server.get('/api/contacts', isAuthForApi, list);
+    server.post('/api/message/sendToUser/:userId', isAuthForApi, messages.sendToUser);
+    server.post('/api/chat/:chatId/message/list', isAuthForApi, messages.list);
```
это не get разве?

maxwell912 replied:
Там же paging

SofiSheromova replied:
ничего не меняет... 
страницы можно (и лучше) в query передавать

maxwell912 replied:
поменял

SofiSheromova commented on server/controllers/messages.ts:
```diff
@@ -0,0 +1,72 @@
+function createMessage(chat, message: string, res: Response) {
+    Message.create({
+        chatId: chat.get('id'),
+        authorId: 1,
```
а как так, а почему

SofiSheromova commented on server/routes.ts:
```diff
@@ -14,10 +16,13 @@ export default (server: Application, passport: PassportStatic, nextApp: NextServ
+    server.post('/api/message/sendToUser/:userId', isAuthForApi, messages.sendToUser);
+    server.post('/api/chat/:chatId/message/list', isAuthForApi, messages.list);
 
+    server.post('/chat', chats.create);
```
**+** нужно уметь по контакту получать чат (обсудили с Глебом)

SofiSheromova replied:
и требуй тут авторизацию как выше

SofiSheromova commented on server/controllers/messages.ts:
```diff
@@ -0,0 +1,72 @@
+        });
+}
+
+export function sendToUser(req: Request , res: Response): void {
```
Не нужно в этой функции искать чат, она уже будет знать id чата
Поиск чата по юзерам должен отдельно быть (с Глебом обсудили это)

SofiSheromova commented on server/controllers/messages.ts:
```diff
@@ -0,0 +1,72 @@
+                if (req.body.message)
+                    createMessage(chat, req.body.message, res);
+                else
+                    res.status(400);
```
Сообщения бы ещё отправлять

maxwell912 replied:
поправил

SofiSheromova reviewed (commented):
Со стилем беды ещё(

Gleb1nho committed:
Added findOrCreate controller for chats

Добавил новую функцию, в основном для удобства. Позволяет искать/создавать чат по двум переданным githubLogin User1&User2

maxwell912 committed:
user-message-api
Добавил код лшибки в ответ
Изменил логику создания чатов

maxwell912 committed:
Merge pull request #21 from urfu-2020/chat-list-for-user

Added findOrCreate controller for chats

maxwell912 committed:
Merge remote-tracking branch 'origin/user-message-api' into user-message-api

# Conflicts:
#	server/controllers/chats.ts
#	server/controllers/messages.ts
#	server/controllers/users.ts

maxwell912 committed:
Merge branch 'develop' into user-message-api

# Conflicts:
#	server/controllers/auth.ts
#	server/controllers/contacts.ts
#	server/models/user.ts

maxwell912 committed:
user-message-api
Стиль

maxwell912 committed:
user-message-api
Стиль

maxwell912 committed:
user-message-api
Нахожу чат по ид а не юзерам

maxwell912 committed:
user-message-api
Добавил роут с чатами текущего юзера

maxwell912 committed:
user-message-api
Изменил тип ссылки в сообщении

maxwell912 committed:
user-message-api
В системе изменил идентификацию юзера с логина на ид

maxwell912 committed:
user-message-api
Добавил роут для списка чатов

maxwell912 committed:
user-message-api
Линтер

maxwell912 committed:
user-message-api
Линтер

maxwell912 committed:
user-message-api
Удалил лишний файл
Отправляю ид в списке пользователей

Gleb1nho reviewed (approved):
Nice!
